### ISSUE TITLE
pca transform should qualify its output: per-sample T²/Q against the model's limits, and RMSEP when the data carries the response

### ISSUE THREAD

**Issue by bitjungle** [bitjungle/gopca#978]
> ### Part of one goal: a GoPCA model should be able to say how far it can be trusted
>
> A prediction is a bare number today. These three turn it into a number with a
> verdict attached — the model carrying its own limits, a command applying them,
> and a deployment acting on them.
>
> | | | |
> |---|---|---|
> | **#977** | the model **carries** its limits | a defect today: the schema promises four fields, `pca analyze` fills them, `pca regress` writes `{}` |
> | **#978** ◀ **this issue** | a command **applies** them | `pca transform` reports T² and Q per sample against those limits, and RMSEP when the data carries the response |
> | **#980** | a deployment **acts** on them | the prediction service — every published value qualified, or not published as a measurement. Scope undecided |
>
> **Each stands alone.** #977 is a defect whatever happens to the others; #978 is
> worth doing for `pca transform` whether or not #980 is ever built; #980 is a
> scope decision before it is an implementation.
>
> The evidence behind all three was measured on `testdata/tecator/`, whose
> extrapolation subsets exist precisely to make a model fail in a known way.

---

*(Merged from #979, which covered the second half. Both change the same output of
the same command, so designing them apart would mean designing it twice.)*

`pca transform` applies a saved model to new data and reports **nothing about how
far it can be trusted**. It prints scores, and predictions when the model carries
a regression block, and that is all — no diagnostics, no comparison against the
model's limits, and no flag to ask for any:

```
$ pca transform --help
      --delimiter   --format   --na-values   --no-headers   --no-index   --output
```

The command already knows this is a gap. It footers every run with:

> Model: 15 components, RMSEC 2.48463, RMSECV 2.87366
> Those figures describe the calibration set. They are not the uncertainty of any
> individual prediction above, and a sample unlike the calibration data can be
> predicted far worse than they suggest.

And `pca regress --help` names the missing figure outright:

> **RMSEP** — an independent test set. **Not produced here**; a test set must be
> kept out of model development entirely.

Two things are needed, and they are one piece of work because they are the same
report.

---

## Part 1 — per-sample: is this sample one the model recognises?

Compute Hotelling's T² and the Q-residual for each projected sample and compare
them against the limits in the model file (#977 puts them there for PCR models;
`pca analyze` already writes them today).

### It works — measured, not assumed

Fitted a model on `testdata/tecator/` **excluding** the extrapolation subsets and
projected them back. Training C+M+T, 215 samples, fat ≤ 49.1%. Limits from the
model file: `T²₉₅ = 27.55`, `Q₉₅ = 1.768e-06`.

| Set | n | median T² | max T² | over T²₉₅ | over Q₉₅ |
|---|---|---|---|---|---|
| C | 129 | 9.18 | 158.6 | 17/129 | 13/129 |
| M | 43 | 8.08 | 97.7 | 4/43 | 5/43 |
| T | 43 | 8.89 | 47.1 | 6/43 | 3/43 |
| **E1** | **8** | **44.79** | **56.2** | **7/8** | **7/8** |
| E2 | 17 | 7.32 | 14.6 | 0/17 | 0/17 |

Sample by sample against the PCR predictions in `testdata/tecator/test_plan.md`:

| Sample | Measured | Predicted | Error | T² | Flagged |
|---|---|---|---|---|---|
| S216 | 54.7 | 49.8 | −4.9 | 35.0 | yes |
| S217 | 55.1 | 63.3 | +8.2 | 48.2 | yes |
| S218 | 56.1 | 54.2 | −1.9 | 44.8 | yes |
| S219 | 56.1 | 53.1 | −3.0 | 25.5 | **no** |
| **S220** | 56.6 | **75.2** | **+18.6** | **56.2** | yes |
| **S221** | 55.1 | **76.9** | **+21.8** | **51.4** | yes |
| S222 | 57.9 | 68.3 | +10.4 | 33.3 | yes |
| S223 | 58.5 | 66.4 | +7.9 | 27.6 | yes |

**The two physically impossible predictions — 75.2% and 76.9% fat, where nothing
in training exceeds 49.1% — carry the two highest T² in the set.** The one sample
not flagged, S219, is among the best predicted.

(The prediction column comes from a leave-one-group-out run trained on C+M+T+E2
rather than C+M+T, so the two models differ slightly. Both exclude E1, so the
comparison is indicative rather than exact.)

---

## Part 2 — per-dataset: how well does this model actually predict?

When the transformed data carries the model's response column, score against it:
RMSEP, bias, SEP, R²pred — the same vocabulary `regress` uses for RMSEC and
RMSECV, so all three are directly comparable.

This is one column away today and is not computed. `pca transform model.json
tecator.csv` reads a file with a populated `Fat#target`, predicts every row,
prints the warning above, and never compares the two.

### The breakdown by Part 1 is where it stops being arithmetic

Model trained on C+M+T, scored on the same file:

| Split | n | RMSEP |
|---|---|---|
| within T²₉₅ | 54 | **2.07** |
| outside T²₉₅ | 14 | **6.50** |
| everything pooled | 68 | *3.48* |

**3.48 describes neither population.** By subset the same data reads T 2.03,
E2 1.96, **E1 8.53**.

"RMSEP 2.07 on the 54 samples I recognise, 6.50 on the 14 I do not" is something
a user can act on. A single number is not — and this is exactly why the two parts
belong in one report rather than two.

(T was in this model's training set, so its own figure understates; E1 and E2 are
genuinely held out. The in/out contrast is the mechanism being illustrated, not a
validation result.)

---

## Two honest caveats, to document rather than hide

**E2 is not flagged, and should not be.** T² and Q describe position in the
*predictor* space. E2 extrapolates in protein — its spectra sit inside the
training cloud while its response relationship does not. Its predictions were
biased (+3.68 mean error) and no X-space diagnostic would say so. **These
statistics screen for novelty in the inputs, not for every way a model can be
wrong**, and the output wording must not imply otherwise.

**The limits are approximations.** 10–13% of the *training* samples exceed their
own 95% limit here, against a nominal 5%. The F-based T² limit and the
Jackson–Mudholkar Q limit both assume conditions real spectra do not meet. A
screen, not a test, and the output should present them as such.

---

## Considered and rejected: per-prediction uncertainty intervals

Recorded so nobody implements this on the assumption that it works.

T² *is* leverage, so an interval looks nearly free:
`ŷ ± 1.96·s·√(1 + 1/n + T²/(n−1))`. Measured on the same model and data:

| | typical sample | worst extrapolation | ratio |
|---|---|---|---|
| interval half-width | 5.18 | 5.92 | **1.14×** |
| actual \|error\| | 0.7 | 15.7 | **10.3×** |

**Five of the eight E1 samples fall outside their own 95% prediction interval.**
The interval widens 14% where the error grows tenfold, because leverage prices
*variance* and this failure is *bias* from model misspecification — the
non-linearity both Tecator papers document. An interval attached to a prediction
of 72.3% fat would be worse than none: a confidence-looking ± on a number wrong
by three times its width.

If intervals are ever added they need this stated beside them, and they must not
displace the T²/Q verdict as the primary safeguard.

---

## Everything needed is already in the artifact

The whole Part 1 table was computed from `train215_pca.json` alone, with no
access to the training data:

* `model.loadings` — variables × components
* `model.explained_variance` — the score variances T² divides by
* `preprocessing.parameters.feature_means` (and stddevs, medians, MADs)
* `model.feature_labels` — for aligning incoming columns by name
* `diagnostics.*` — after #977

Per sample: preprocess, project, `T² = Σ tₐ²/λₐ`, `Q = ‖xᶜ − t·Pᵀ‖²`, compare. No
new mathematics and no new stored state.

## Not part of this: variable alignment already works

Checked, since everything here depends on it. `pca transform` aligns incoming
columns **by name**: reversing the order of all 100 spectral columns produces
predictions identical to the correct file, and a renamed or missing column is
refused with `missing required features in data: [850.0]`. No work needed.

---

## Shape

Deliberately not prescribing the interface. The pieces seem to be:

* per-sample `t2` and `q` alongside the existing scores and predictions, in both
  the table and the JSON output;
* a comparison against the stored limits expressed as something a reader cannot
  miss — a marked row, or an explicit `within_model` verdict;
* RMSEP, bias, SEP, R²pred when a response column is present; absent, behave
  exactly as now;
* those figures **split by the verdict**, and by any categorical column the user
  nominates — `Set#category` on Tecator gives the C/M/T/E1/E2 breakdown directly,
  which is what that dataset is designed for;
* a summary line saying how many of N fell outside, since that is the number a
  user acts on;
* sensible behaviour when the model carries no limits — one written before #977,
  or one where the Q limit is legitimately absent because every component was
  retained. Silence is wrong; so is refusing;
* say plainly when the scored data was part of the model's training set if that
  can be detected, and that it cannot be if it cannot — an RMSEP computed on
  training data is the classic way this figure misleads.

**Whether `transform` should ever refuse to print a prediction:** I do not think
so. Suppressing the number hides information the user may need. Qualifying it
loudly is the better default.

## Dependencies and why now

Needs #977 for PCR models. Works today for models from `pca analyze`, which
already carry populated limits — so the per-sample half is not strictly blocked.

`testdata/tecator/test_plan.md` cannot currently reproduce the published
benchmark figures — 2.78 for a linear model, 2.86 PLS, 2.92 PCR — because those
are all SEP on the T subset after tuning on C+M, and nothing in the suite scores
a saved model against a held-out file. This closes that, and turns the dataset's
five-subset design from documentation into something the tool exercises.

**Comment by bitjungle** [comment]
Schema check, prompted by the question of whether any of this needs a v3.

## It does not. But the naming is already decided, and I had not noticed

`schemas/v2/results.schema.json` **already models per-sample diagnostics**, and has since before this issue:

| field | description in the schema |
|---|---|
| `hotelling_t2` | Hotelling's T-squared statistic for each sample |
| `mahalanobis` | Mahalanobis distance for each sample |
| `rss` | Residual sum of squares for each sample |
| `is_outlier` | Outlier flag for each sample |

`pca analyze --include-metrics` fills all four. So the per-sample half of this issue is not inventing a representation — **it is computing, for new data, the same four quantities the model file already carries for the training data.**

That matters for three decisions I left open in the body:

**1. Name the residual `rss`, not `q_residual`.** The Q-residual and the residual sum of squares are the same quantity, and the schema has already chosen a word. Introducing a second name for it in `transform`'s output would mean one concept spelled two ways in one project — the defect #973 was about. If `q_residual` is the better name, rename it *everywhere*, in a v3, deliberately; do not add it alongside.

**2. `is_outlier` already exists as a verdict field.** Whatever "within the model" verdict this issue adds should either be that field or be clearly distinguished from it. `is_outlier` in the training context means "exceeds a limit fitted on data including this sample"; for new data it would mean "exceeds a limit fitted without it". Same name, materially different claim — worth being explicit rather than reusing it silently.

**3. `mahalanobis` sits alongside `hotelling_t2`.** For a PCA model with all components retained the two coincide; with a truncated model they do not. If `transform` reports T² it should probably report Mahalanobis too, or say why not, since the model file offers both for the training rows and a reader will compare.

## Nothing here needs a schema change

The fields exist. What `transform` emits is a *report*, not a model artifact, and is not schema-governed today — so the freedom is real, and the argument above is about consistency rather than validity.

The one genuine schema question belongs to #977, not here: how to express a Q limit that legitimately does not exist, since `0` is a valid limit value and `omitempty` currently renders both as absent. Whatever is decided there, this issue has to render it.

## Also verified

`schemas/` and `pkg/validation/schemas/` are in sync across v1 and v2, file for file — a previous note in this project had recorded them drifting, so it was worth confirming before proposing anything that touches them.

**Comment by bitjungle** [comment]
**Correction to the E2 caveat in the body.** I overstated it, and the correction makes the diagnostics look *better*, not worse.

The body says of E2: *"Its predictions were biased (+3.68 mean error) and no X-space diagnostic would say so"* — presenting it as a failure the flag misses. I quoted that bias without its comparator, and from a different model than the one in the RMSEP table two sections above.

Both figures are valid held-out measurements of E2. Side by side:

| | model | T | E1 | **E2** |
|---|---|---|---|---|
| RMSE | leave-one-group-out, 5 comps | 3.66 | 11.72 | **4.17** |
| bias | | −0.68 | +7.15 | **+3.68** |
| RMSEP | C+M+T, 16 comps | 2.03 | 8.53 | **1.96** |
| bias | | −0.14 | +4.55 | **+1.28** |

**E2 is predicted about as well as T under both models** — marginally worse under one, marginally better under the other. It is not an example of an unflagged failure, and not flagging it is the correct outcome rather than a miss.

So on everything actually measured here, the diagnostics behaved correctly in both directions: they flagged 7 of the 8 samples that were badly predicted, did not flag the one E1 sample that was predicted well (S219), and did not flag the 17 E2 samples that did not need it.

The underlying limitation still holds and still belongs in the caveats, but as a property rather than a demonstration: **T² and Q describe position in the predictor space, so a change in the *response* relationship — the same spectra meaning something different — would not be visible to them.** Tecator happens not to contain such a case. That is worth stating precisely because it is the failure mode nothing in this dataset can rehearse, and anything built on top of these diagnostics needs a different mechanism for it: periodic reference measurements, which is what the RMSEP half of this issue provides the arithmetic for.

**Comment by bitjungle** [comment]
Cross-reference: **#980** proposes the prediction service this issue is a prerequisite for — a separate binary that loads a regression model and predicts from live data, MQTT first.

Nothing here depends on that, and this issue is worth doing whether or not #980 is ever built: `pca transform` has the same gap today. But #980 needs the per-sample computation and the limits comparison to exist somewhere tested, and here is the right place for them — one implementation rather than two.
